framework/math/NonLinearSolver/NonLinearSolverOptions.h

```diff
@@ -0,0 +1,39 @@
+#ifndef CHITECH_NONLINEARSOLVEROPTIONS_H
+#define CHITECH_NONLINEARSOLVEROPTIONS_H
+
+#include "ChiObject.h"
+
+namespace chi_math
+{
+
+class NonLinearSolverOptions : public ChiObject
+{
+public:
+  static chi::InputParameters GetInputParameters();
+  explicit NonLinearSolverOptions(const chi::InputParameters& params);
+  NonLinearSolverOptions() = default;
+
+  std::string nl_method_ = "JFNK";
+  std::string l_method_ = "gmres";
+
+  chi::ParameterBlock pc_options_;
+
+  std::string petsc_snes_type_ = "newtonls";
+
+  double nl_rel_tol_ = 1.0e-8;
+  double nl_abs_tol_ = 1.0e-8;
+  double nl_sol_tol_ = 1.0e-50;
+  int    nl_max_its_ = 50;
+  int    nl_max_r_evaluations_ = -1;
+  int    l_max_failed_iterations_ = 1000;
+  double l_rel_tol_ = 1.0e-8;
+  double l_abs_tol_ = 1.0e-8;
+  double l_div_tol_ = 1.0e6;
+  int    l_max_its_ = 100;
+  int    l_gmres_restart_intvl_ = 30;
+  double l_gmres_breakdown_tol_ = 1.0e6;
```

Comment:
I have that of something like this. It becomes tricky when you want it to appear completely pluggable because PETSc is so different from other design patterns.

We can easily define LinearSolverOptions and then make NonLinearSolverOptions inherit from it. However, that would be a scope for a different PR.

Reply:
This may merit more discussion offline as that I'm not sure what you really mean by "completely pluggable".